### ISSUE TITLE
Add Check Support for Java 17 Sealed Classes: New Check SealedClassesShouldPermitsSubClasses

### ISSUE THREAD

**Issue by mahfouz72** [checkstyle/checkstyle#14975]
child of #14969

> the sealed class may omit the permits clause and the Java compiler will infer the permitted subclasses from the declarations in the source file. (The subclasses may be auxiliary or nested classes.) 

A sealed class restricts which other classes can extend it. This is typically achieved using the permits clause.
However, if a sealed class does not specify any permitted subclasses using the permits clause, it implicitly allows any class within the same file to extend it. This can lead to unintended inheritance classes that should not have been allowed to extend the sealed class can do so at any time. 

I propose the implementation of a check that violates any sealed class or interface without a `permits` clause. This check will help ensure that we explicitly defining permitted subclasses, therefore avoiding unintended extensions.

--- 
**Config**

```XML
$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="PatternVariableAssignment"/>
    </module>
</module>
```

---
**Example**

```Java
public class SealedClasses {
     sealed class A {} // violation
     final class B extends A {}
     final class C extends A {}
     final class D { }    // this is not intended to be a subclass of A but it can do so anytime as there is no restrictive permits clause
 }
```
---
**Expected Output**

```
$ java -jar checkstyle-10.19.0-all.jar -c config.xml SealedClasses.java 
Starting audit...
[ERROR] D:\SealedClasses.java:2:5: subclasses are missing from permits clause of a sealed class
Audit done.
Checkstyle ends with 1 errors.
```

**Comment by rnveach** [comment]
Can you show what valid code should be (violations corrected by user).

> ShouldPermitsSubClasses
> subclasses are missing from permits clause of a sealed class
> violates any sealed class or interface without a permits clause

The first 2 sentences seem to conflict with the last sentence in the fact of sub-classes versus regular classes. What if there are no sub-classes in the file, will it still be a violation?

Also, we talked in discord about a `SealedClassShouldHaveExplicitPermitsList` check. Would it make more sense to just implement that type of check, where we violate all seals without a permit list?

**Comment by mahfouz72** [comment]
> Can you show what valid code should be (violations corrected by user).

```
public class SealedClasses {
     sealed class A permits B, C {}
     final class B extends A {}
     final class C extends A {}
     final class D { }  
 }
```
> Also, what if we have another input file where sealed class A isn't a sub-class.

can you clarify this because sealed class A here isn't a sub-class?

> What if there are no sub-classes in the file, will it still be a violation?

this won't compile => `sealed class must have subclasses`

> Also, we talked in discord about a SealedClassShouldHaveExplicitPermitsList check. Would it make more sense to just implement that type of check, where we violate all seals without a permit list?

yes,  that is true the check name now may be confusing. I will edit it to keep everything consistent. we will violate seals without permits

**Comment by rnveach** [comment]
> Also, what if we have another input file where sealed class A isn't a sub-class.
> can you clarify this because sealed class A here isn't a sub-class?

I consider A a sub-class because it is inside `SealedClasses`. Maybe the proper term is really nested.

> this won't compile => sealed class must have subclasses

So a top-level class can never be sealed as I understand it?

**Comment by mahfouz72** [comment]
It can be top-level. I think you want to see an example like this: 

```Java
public sealed class Test {
    public final class M extends Test { }
    public final class X extends Test { }
    public final class Y { }
}

```

the above example is valid after fixing the violation it will be 

```Java
public sealed class Test permits Test.M, Test.X {
    public final class M extends Test2 { }
    public final class X extends Test2 { }
    public final class Y { }
}
```

so you can say that all nested classes inside top-level sealed classes without `permits` can extend these top-level classes. with `permits` we can restrict the extension to the classes in the `permits list only
`

**Comment by nrmancuso** [comment]
I am good to approve this issue if we can update the check's message to `Sealed classes or interfaces should explicitly declare permitted subclasses.`

**Comment by nrmancuso** [comment]
@rnveach please add approved label when you are good

**Comment by nrmancuso** [comment]
This check can live in with design checks.
